javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -655,19 +676,14 @@ private Vector(int indexShift, HashArrayMappedTrie<Integer, T> trie) {
     @Override
     public Vector<T> filter(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
-        HashArrayMappedTrie<Integer, T> filtered = HashArrayMappedTrie.empty();
-        for (T t : this) {
-            if (predicate.test(t)) {
-                filtered = filtered.put(filtered.size(), t);
-            }
-        }
 
-        if (filtered.isEmpty()) {
-            return Vector.empty();
-        } else if (filtered.size() == size()) {
+        final Vector<T> results = ofAll(iterator().filter(predicate));
+        if (results.isEmpty()) {
```

Comment:
This is redundant. If `results.isEmpty()` then we have empty singleton and can return `results`